opm/simulators/wells/GasLiftSingleWellGeneric.cpp

```diff
@@ -484,44 +501,152 @@ std::pair<double, bool>
+        }
+    }
+    return { new_gas_rate, /*limit=*/false};
 }
 
-std::tuple<double,double,bool,bool>
+std::pair<double, bool>
+GasLiftSingleWellGeneric::
+getWaterRateWithGroupLimit_(const double new_water_rate, const double water_rate) const
+{
+    const double delta_water = new_water_rate - water_rate;
+    const auto &pairs =
+        this->group_info_.getWellGroups(this->well_name_);
+    for (const auto &[group_name, efficiency] : pairs) {
+        auto gr_water_target_opt = this->group_info_.waterTarget(group_name);
+        if (gr_water_target_opt) {
+            double gr_water_rate =
+                this->group_info_.waterRate(group_name);
+            double new_gr_water_rate = gr_water_rate + efficiency * delta_water;
+            if (new_gr_water_rate > *gr_water_target_opt) {
+                const std::string msg = fmt::format("limiting water rate to group target: "
+                    "computed group rate: {}, target: {}", new_gr_water_rate, *gr_water_target_opt);
+                displayDebugMessage_(msg);
+                double new_rate = water_rate + (*gr_water_target_opt - gr_water_rate) / efficiency;
+                return { std::min(new_rate, new_water_rate), /*limit=*/true};
+            }
+        }
+    }
+    return { new_water_rate, /*limit=*/false};
+}
+
+std::tuple<double, double, bool, bool>
+GasLiftSingleWellGeneric::
+getLiquidRateWithGroupLimit_(const double new_oil_rate, const double oil_rate,
+                             const double new_water_rate, const double water_rate) const
+{
+    const double delta_water = new_water_rate - water_rate;
+    const double delta_oil = new_oil_rate - oil_rate;
+    const auto &pairs =
+        this->group_info_.getWellGroups(this->well_name_);
+    for (const auto &[group_name, efficiency] : pairs) {
+        auto gr_liquid_target_opt = this->group_info_.liquidTarget(group_name);
+        if (gr_liquid_target_opt) {
+            double gr_water_rate =
+                this->group_info_.waterRate(group_name);
+            double gr_oil_rate =
+                this->group_info_.oilRate(group_name);
+            double new_gr_water_rate = gr_water_rate + efficiency * delta_water;
+            double new_gr_oil_rate = gr_oil_rate + efficiency * delta_oil;
+            double new_gr_liquid_rate = new_gr_water_rate + new_gr_oil_rate;
+            if (new_gr_liquid_rate > *gr_liquid_target_opt) {
+                const std::string msg = fmt::format("limiting liquid rate to group target: "
+                    "computed group rate: {}, target: {}", new_gr_liquid_rate, *gr_liquid_target_opt);
+                displayDebugMessage_(msg);
+                double oil_fraction = new_gr_oil_rate / new_gr_liquid_rate;
+                double water_rate_limited = water_rate + (1.0 - oil_fraction) * (new_gr_liquid_rate - *gr_liquid_target_opt) / efficiency;
+                double oil_rate_limited = oil_rate + oil_fraction * (new_gr_liquid_rate - *gr_liquid_target_opt) / efficiency;
```

Comment:
This should probably be
```
double water_rate_limited = water_rate + (1.0 - oil_fraction) * (*gr_liquid_target_opt - gr_liquid_rate) / efficiency;
double oil_rate_limited = oil_rate + oil_fraction * (*gr_liquid_target_opt - gr_liquid_rate) / efficiency;
```
where `gr_liquid_rate = gr_water_rate + gr_oil_rate` since we are reducing the liquid rate down to the target.